### PR TITLE
New chunk should be equal/larger than the one it replaces

### LINKED ISSUE #17205: Failure: StringBuilder.set_Length: The new chunk should be larger than the one it is replacing.
For `Windows_NT x64 Checked corefx_baseline` and `Ubuntu x64 Checked corefx_baseline` https://github.com/dotnet/coreclr/pull/17204 also https://github.com/dotnet/coreclr/pull/17196#issuecomment-375945569

Ubuntu x64 Checked Build and Test (Jit - CoreFx) failure:
```
Discovering: System.Security.SecureString.Tests
Discovered:  System.Security.SecureString.Tests
Starting:    System.Security.SecureString.Tests
Assertion Failed
The new chunk should be larger than the one it is replacing.

   at System.Text.StringBuilder.set_Length(Int32 value)
   at System.Text.StringBuilder.Remove(Int32 startIndex, Int32 length)
   at System.Security.Tests.SecureStringTests.GrowAndContract_Small() in /mnt/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_ubuntu_corefx_baseline_prtest/_/fx/src/System.Security.SecureString/tests/SecureStringTests.cs:line 426
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass46_1.<<InvokeTestMethodAsync>b__1>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at ...
```

Perhaps related to https://github.com/dotnet/coreclr/pull/16926?

/cc @maryamariyan @danmosemsft @jkotas

### PR DESCRIPTION
**Pull request by maryamariyan**
Fixes #17205

In PR dotnet/coreclr#16926 we changed the behavior on a non-empty StringBuilder such that it will **not** always grow the last chunk array length on `sb.Length=value` operation where value is smaller than current Length.

The goal for this change was to keep Capacity within finite range as we loop multiple times to Clear and Append and Insert into the StringBuilder.

This PR removes the assertion that makes sure we always grow it's last chunk array.

cc: @jkotas @danmosemsft  @benaadams

### CODE REVIEW

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -486,12 +486,18 @@ public int Length
                         // Avoid possible infinite capacity growth.  See https://github.com/dotnet/coreclr/pull/16926
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: this comment is now wrong?
```
                        // We crossed a chunk boundary when reducing the Length. We must replace this middle-chunk with a new larger chunk,
                        // to ensure the original capacity is preserved.
```

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 2]
```diff
@@ -486,12 +486,18 @@ public int Length
...
+                        if (newLen > chunk.m_ChunkChars.Length)
+                        {
+                            char[] newArray = new char[newLen];
+                            Array.Copy(chunk.m_ChunkChars, 0, newArray, 0, chunk.m_ChunkLength);
+                            m_ChunkChars = newArray;
+                        }
+                        else
+                        {
+                            Debug.Assert(newLen == chunk.m_ChunkChars.Length, "The new chunk should be larger or equal to the one it is replacing.");
+                            m_ChunkChars = chunk.m_ChunkChars;
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]:
```
// Special case where the capacity we want to keep corresponds exactly to the size of the content.
// Just take ownership of the array.
```

Not sure the assert is worth having now.

Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Can you check there's a test that goes down this path?

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 3]
```diff
@@ -486,12 +486,18 @@ public int Length
...
+                        }
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: Is it worth asserting at the end of the method that Length == value

**Comment by danmosemsft** [comment]
Can you please add specific to StringBuilder that fails before your change and passes after? We should not rely on other test libraries for coverage necessary to protect our functionality.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 1 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: I'll move this to the if block.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 2 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: I'll move this comment into the else block

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 3 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Added this assert to the end:

`Debug.Assert(Length == value, "Something went wrong setting Length.");`

**Comment by maryamariyan** [comment]
Will be updating test to corefx PR dotnet/corefx#28038

File: vim.exe.stackdump [hunk 4]
```diff
@@ -0,0 +1,19 @@
+Stack trace:
+Frame        Function    Args
+00180238000  0018005D19E (00180223639, 00180223C39, 001802342F0, 000FFFFB720)
```
Comment (danmosemsft) [inline vim.exe.stackdump]: Needs license header 😸

Comment (danmosemsft) [reply vim.exe.stackdump]: Actually you could add `.exe.stackdump` to our `.gitignore` at the root.

Comment (maryamariyan) [reply vim.exe.stackdump]: oh no! didn't mean to add this :)

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 5]
```diff
@@ -480,24 +480,32 @@ public int Length
...
                         // Avoid possible infinite capacity growth.  See https://github.com/dotnet/coreclr/pull/16926
                         int capacityToPreserve = Math.Min(Capacity, Math.Max(Length * 6 / 5, m_ChunkChars.Length));
                         int newLen = capacityToPreserve - chunk.m_ChunkOffset;
-                        char[] newArray = new char[newLen];
-
-                        Debug.Assert(newLen > chunk.m_ChunkChars.Length, "The new chunk should be larger than the one it is replacing.");
-                        Array.Copy(chunk.m_ChunkChars, 0, newArray, 0, chunk.m_ChunkLength);
+                        if (newLen > chunk.m_ChunkChars.Length)
+                        {
+                            // We crossed a chunk boundary when reducing the Length. We must replace this middle-chunk with a new larger chunk,
```
[the remark this answers is not in the document]
Comment (maryamariyan) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: updated

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by maryamariyan** [comment]
### About the Use Case:
The code snippet below is taken from failing test: `System.Security.Tests.SecureStringTests.GrowAndContract_Small`.
The last statement in the code snippet below will exercise the code path in this PR.

```
            var rand = new Random(42);
            var sb = new StringBuilder(string.Empty);
            for (int i = 0; i < 100; i++)
            {
                char c = (char)('a' + rand.Next(0, 26));
                int addPos = rand.Next(0, sb.Length);
                sb.Insert(addPos, c);
            }
            while (sb.Length > 1)
            {
                int removePos = rand.Next(0, sb.Length);
                sb.Remove(removePos, 1);
            }
            sb.Clear(); //sb.Length = 0;//sb.Remove(0, 1);
```

This will result in a StringBuilder with with length of 1 and two chunks. 
The first chunk has offset=0, length=1, array length=16, and the second chunk has offset=1 and length=0,  array length=16. 
The Capacity is 17. (=last chunk array size, 16 + last chunk offset, 1).

In this case, it should be OK to reduce the Capacity value to 16 as we set Length to 0 and keep chunk array length the same value, rather than expanding the chunk length to 17 as we were doing prior to the change in PR dotnet/coreclr#16926

**Comment by benaadams** [comment]
Failures seem are infrastructure:
* Ubuntu x64 Checked Build and Test (Jit - CoreFx)  - time out during build 
* Tizen - usual
* OSX10.12 x64 - downstream failure

Windows_NT x64 Checked Build and Test (Jit - CoreFx) - passes which was failing on this code before

**Pull request merged**